File: code/done/977.squares-of-a-sorted-array.go
/*
 * @lc app=leetcode id=977 lang=golang
 *
 * [977] Squares of a Sorted Array
 */

package main

// @lc code=start
func sortedSquares(nums []int) []int {
	length := len(nums)
	result := make([]int, length)
	left := 0
	right := length - 1
	k := length - 1
	for left <= right {
		if nums[left]*nums[left] > nums[right]*nums[right] {
			result[k] = nums[left] * nums[left]
			left++
		} else {
			result[k] = nums[right] * nums[right]
			right--
		}
		k--
	}
	return result
}

// @lc code=end

File: code/done/977.squares-of-a-sorted-array.py
#
# @lc app=leetcode id=977 lang=python3
#
# [977] Squares of a Sorted Array
#

# @lc code=start
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        left = 0
        length = len(nums)
        right = length - 1
        result = [0]*length
        for i in range(length-1, -1, -1):
            if nums[left]**2 >= nums[right]**2:
                result[i] = nums[left]**2
                left += 1
            else:
                result[i] = nums[right]**2
                right -= 1
        return result


# @lc code=end
